### PR TITLE
Implement XEP-0004: Data Forms in a separate module

### PR DESCRIPTION
**Pull request by chrzaszcz**
The main goal is to have a single module that implements [XEP-0004](https://xmpp.org/extensions/xep-0004.html) and [XEP-0068](https://xmpp.org/extensions/xep-0068.html) for better separation of concerns, reusability and less `xmlel` manipulation.

The implementation is divided into two parts:
  - Parsing of submitted forms into a `parsed_form()` map
  - Form construction, which builds XML from a `form()` map

The two data structures are different, because they serve different purposes, but they are both maps.

The form should be parsed/constructed only once. Ideally we should minimize the operations on plain XML elements, i.e.:
  - Parsing should be done as early as possible.
  - Construction should be done as late as possible.

Other changes:
  - Fix a bug in event pusher that was randomly triggered by the changes
  - Improve error messages

Excluded from this PR:
  - Improving test coverage. There are some lines that were not covered before, and can be covered by testing error conditions with missing or malformed data forms. We can do it in a later PR, but here I wanted to limit the change set.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/176278) / small_tests / 71a170cc9ce60b15e5afded3bb4ae92c8e972aef
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4028/176278/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4028/176278/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-05-25_09.47.20/lib.mongooseim.logs/run.2023-05-25_09.47.20/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/176331) / small_tests / f1c34e23cebf2b91c262081cf20ea062202351e5
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4028/176331/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4028/176331/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-05-25_10.30.17/lib.mongooseim.logs/run.2023-05-25_10.30.18/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/176386) / small_tests / e37cc5a80f9243ee6ebbfd87c2b858970b4453d4
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4028/176386/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4028/176386/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-05-25_10.45.49/lib.mongooseim.logs/run.2023-05-25_10.45.50/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/176415) / small_tests / 2e878052d2304e9cecb442c39b1718fd9240a47c
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4028/176415/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4028/176415/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-05-25_12.31.11/lib.mongooseim.logs/run.2023-05-25_12.31.12/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/176530) / small_tests / c9ab439221fc49acaa06ca203740d40c3717ac66
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4028/176530/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4028/176530/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-05-26_09.45.29/lib.mongooseim.logs/run.2023-05-26_09.45.29/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/176559) / small_tests / 46b403d7da89c4b67f84d12c208eb7e94ed42608
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4028/176559/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4028/176559/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-05-26_13.50.19/lib.mongooseim.logs/run.2023-05-26_13.50.19/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/176583) / small_tests / 1f688d17cce44c3a175c47d6cb593f6ea6f9db94
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4028/176583/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4028/176583/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-05-26_14.13.23/lib.mongooseim.logs/run.2023-05-26_14.13.23/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/176630) / small_tests / c9624c162d77fa721f106b9f3dd409e8c08df951
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4028/176630/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4028/176630/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-05-26_16.54.59/lib.mongooseim.logs/run.2023-05-26_16.54.59/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/176844) / small_tests / 3c3403143772c4a2332fa3c8ea77b4074cf85745
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4028/176844/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4028/176844/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-05-30_14.18.28/lib.mongooseim.logs/run.2023-05-30_14.18.29/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/176848) / small_tests / 4ff9d5a9b3611150026c9d37ce5b59261af05781
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4028/176848/small_tests.24.3.4.9-amd64) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4028/176848/small_tests.24.3.4.9-amd64/small/ct_run.mongooseim@localhost.2023-05-30_14.19.42/lib.mongooseim.logs/run.2023-05-30_14.19.42/suite.log.html)

File: src/mod_muc_room.erl [hunk 1]
```diff
@@ -3221,26 +3223,26 @@ process_authorized_iq_owner(From, get, Lang, SubEl, StateData, _StateName) ->
...
         false -> {error, mongoose_xmpp_errors:not_acceptable(<<"en">>, <<"not allowed to configure">>)}
     end.
 
--spec is_allowed_log_change(exml:element(), state(), jid:jid()) -> boolean().
-is_allowed_log_change(XEl, StateData, From) ->
-    case lists:keymember(<<"muc#roomconfig_enablelogging">>, 1,
-             jlib:parse_xdata_submit(XEl)) of
+-spec is_allowed_log_change([{binary(), [binary()]}], state(), jid:jid()) -> boolean().
+is_allowed_log_change(XData, StateData, From) ->
+    case lists:keymember(<<"muc#roomconfig_enablelogging">>, 1, XData) of
```
Comment (JanuszJakubiec) [inline src/mod_muc_room.erl:3245]: Wrong indentation, minor

File: src/mod_muc_room.erl [hunk 2]
```diff
@@ -3250,10 +3252,9 @@ is_allowed_log_change(XEl, StateData, From) ->
     end.
 
 
--spec is_allowed_persistent_change(exml:element(), state(), jid:jid()) -> boolean().
-is_allowed_persistent_change(XEl, StateData, From) ->
-    case lists:keymember(<<"muc#roomconfig_persistentroom">>, 1,
-             jlib:parse_xdata_submit(XEl)) of
+-spec is_allowed_persistent_change([{binary(), [binary()]}], state(), jid:jid()) -> boolean().
+is_allowed_persistent_change(XData, StateData, From) ->
+    case lists:keymember(<<"muc#roomconfig_persistentroom">>, 1, XData) of
```
Comment (JanuszJakubiec) [inline src/mod_muc_room.erl:3257]: Wrong indentation, minor

File: src/mod_muc_room.erl [hunk 3]
```diff
@@ -3221,26 +3223,26 @@ process_authorized_iq_owner(From, get, Lang, SubEl, StateData, _StateName) ->
...
     end.
 
--spec is_allowed_log_change(exml:element(), state(), jid:jid()) -> boolean().
-is_allowed_log_change(XEl, StateData, From) ->
-    case lists:keymember(<<"muc#roomconfig_enablelogging">>, 1,
-             jlib:parse_xdata_submit(XEl)) of
+-spec is_allowed_log_change([{binary(), [binary()]}], state(), jid:jid()) -> boolean().
+is_allowed_log_change(XData, StateData, From) ->
+    case lists:keymember(<<"muc#roomconfig_enablelogging">>, 1, XData) of
     false ->
```
Comment (JanuszJakubiec) [inline src/mod_muc_room.erl:3246]: Wrong indendation, minor

File: src/mod_muc_room.erl [hunk 4]
```diff
@@ -3250,10 +3252,9 @@ is_allowed_log_change(XEl, StateData, From) ->
...
 
 
--spec is_allowed_persistent_change(exml:element(), state(), jid:jid()) -> boolean().
-is_allowed_persistent_change(XEl, StateData, From) ->
-    case lists:keymember(<<"muc#roomconfig_persistentroom">>, 1,
-             jlib:parse_xdata_submit(XEl)) of
+-spec is_allowed_persistent_change([{binary(), [binary()]}], state(), jid:jid()) -> boolean().
+is_allowed_persistent_change(XData, StateData, From) ->
+    case lists:keymember(<<"muc#roomconfig_persistentroom">>, 1, XData) of
     false ->
```
Comment (JanuszJakubiec) [inline src/mod_muc_room.erl:3258]: Wrong indentation, minor

**Pull request merged**